toxcore/query.c

```diff
@@ -0,0 +1,347 @@
+
+    memcpy(built + size, our_public_key, crypto_box_PUBLICKEYBYTES);
+    size += crypto_box_PUBLICKEYBYTES;
+
+    memcpy(built + size, nonce, crypto_box_NONCEBYTES);
+    size += crypto_box_NONCEBYTES;
+
+    memcpy(built + size, encrypted, status);
+
+    return 1 + crypto_box_PUBLICKEYBYTES + crypto_box_NONCEBYTES + status;
+}
+
+static int q_send(DHT *dht, P_QUERY send)
+{
+    uint8_t packet[QUERY_PKT_ENCRYPTED_SIZE(send.length)];
+
+    size_t written_size = q_build_pkt(send.key, dht->self_public_key, dht->self_secret_key, NET_PACKET_DATA_REQUEST,
+                                      send.name, send.length, packet);
+    // TODO(grayhatter) add tox_assert(written_size == QUERY_PKT_ENCRYPTED_SIZE(send.length));
+
+    if (written_size != -1) {
+        return sendpacket(dht->net, send.ipp, packet, written_size) != written_size;
+    }
+
+    return -1;
+}
+
+static P_QUERY q_make(IP_Port ipp, const uint8_t key[TOX_PUBLIC_KEY_SIZE], const uint8_t *name, size_t length)
+{
+    P_QUERY new;
+    memset(&new, 0, sizeof(P_QUERY));
+
+    new.ipp = ipp;
+    id_copy(new.key, key);
+    new.name = calloc(1, length);
+    memcpy(new.name, name, length);
+    new.length = length;
+
+    // new.query_nonce = random_64b(); // TODO(grayhatter) readd nonce
+
+    new.tries_remaining = 2;
+    new.next_timeout = unix_time() + QUERY_TIMEOUT;
+    return new;
+}
+
+/** query_send_request
+ *
+ *
+ */
+int query_send_request(Tox *tox, const char *address, uint16_t port, const uint8_t *key,
+                       const uint8_t *name, size_t length)
+{
+    Messenger *m = (Messenger *)tox;
+
+    struct addrinfo *info, *root;
+
+    if (getaddrinfo((char *)address, NULL, NULL, &root) != 0) {
+        return -1;
+    }
+
+    IP_Port ipp;
+    memset(&ipp, 0, sizeof(IP_Port));
+    ipp.port = htons(port);
+
+    info = root;
+
+    do {
+        if (info->ai_socktype && info->ai_socktype != SOCK_DGRAM) {
+            continue;
+        }
+
+        ipp.ip.family = info->ai_family;
+
+        if (info->ai_family == AF_INET) {
+            ipp.ip.ip4.in_addr = ((struct sockaddr_in *)info->ai_addr)->sin_addr;
+            break;
+        } else if (info->ai_family == AF_INET6) {
+            ipp.ip.ip6.in6_addr = ((struct sockaddr_in6 *)info->ai_addr)->sin6_addr;
+            break;
+        } else {
+            continue;
+        }
+    } while ((info = info->ai_next));
+
+    freeaddrinfo(root);
+
+    if (info == NULL) {
+        return -1; // No host found
+    }
+
+    P_QUERY new = q_make(ipp, key, name, length);
+
+    // Verify name isn't currently pending response
+    if (q_check(m->dht->queries, new, 1) != -1) {
+        return -2;
+    }
+
+    // Send request
+    if (q_send(m->dht, new) == 0) {
+        if (q_add(m->dht->queries, new)) {
+            return 0;
+        }
+
+        return -3;
+    }
+
+    return -4;
+}
+
+int query_handle_toxid_response(void *object, IP_Port source, const uint8_t *pkt, uint16_t length, void *userdata)
```

Comment:
@iphydf you're not missing a commit, I submitted all my response and decided to write one more test. When I did, I started getting segfaults. The problem was network returns an IP_Port struct, not a pointer.